tket/src/Circuit/include/Circuit/Boxes.hpp

```diff
@@ -708,4 +708,58 @@ class StabiliserAssertionBox : public Box {
   mutable std::vector<bool> expected_readouts_;
 };
 
+class ToffoliBox : public Box {
+ public:
+  typedef std::vector<std::vector<bool>> cycle_permutation_t;
+
+  struct transposition_t {
+    std::vector<bool> first;
+    std::vector<bool> middle;
+    std::vector<bool> last;
+  };
+
+  typedef std::vector<transposition_t> cycle_transposition_t;
+
+  typedef std::vector<std::pair<std::vector<bool>, unsigned>> gray_code_t;
+
+  /**
+   * Construct from a map between input and output basis states.
+   * Map entries should produce a cycle, i.e. if A maps to B but B
+   * B has no entry then it will throw an invalid_argument error.
+   * Any basis state not in a permutation cycle will be assumed to
+   * take the identity.
+   * If each basis state is not the same size will throw an
+   * invalid_argument error.
+   *
+   * @param _permutation map between basis states
+   * @param reorder if true does additional optimisation to reudce CnX count
+   */
+  explicit ToffoliBox(
```

Comment:
I think this needs to take the number of qubits as a parameter. Otherwise with an empty vector (which could well arise programmatically) we don't know how many qubits the circuit should have, (Then the constructor should check that the lengths of all vector-of-bools match the stated number of qubits.)

Reply:
Done.